web3/utils/datatypes.py

```diff
@@ -0,0 +1,50 @@
+from functools import (
+    partial,
+)
+from eth_utils import (
+    to_dict,
+)
+from toolz.itertoolz import (
+    concat
+)
+
+
+def verify_attr(class_name, key, base):
+    if not hasattr(base, key):
+        raise AttributeError(
+            "Property {0} not found on {1} class. "
+            "`{1}.factory` only accepts keyword arguments which are "
+            "present on the {1} class".format(key, class_name)
+        )
+
+
+class PropertyCheckingFactory(type):
+    @to_dict
+    def normalize_properties(attributes, values, normalizers):
+        for attribute, value in zip(attributes, values):
+            if attribute in normalizers:
+                normalizer = normalizers[attribute]
+                yield attribute, normalizer(value)
+            else:
+                yield attribute, value
+
+    def __init__(cls, name, bases, namespace, **kargs):
+        # see PEP487.  To accept kwargs in __new__, they need to be
+        # filtered out here.
+        super().__init__(name, bases, namespace)
+
+    def __new__(mcs, name, bases, namespace, normalizers={}):
```

Comment:
My general rule is that mutable data structures are never ok as default arguments for functions.

```
In [1]: def f(a, x=[]):
   ...:     x.append(a)
   ...:     return x
   ...:

In [2]: f(1)
Out[2]: [1]

In [3]: f(2)
Out[3]: [1, 2]

In [4]: f(3)
Out[4]: [1, 2, 3]
```

This specific case doesn't appear to have these problems, but I still think it's a good practice for us to avoid this pattern in favor of the following.

```python
def __new__(mcs, name, bases, namespace, normalizers=None):
    if normalizers is None:
        normalizers = {}
    ...
```